mGCA: Syntactic type mismatch diagnostic can leak `{type error}` for forbidden inferred types `_` in the type of type-level const items

With *syntactic* type mismatch I'm referring to mismatches detected by the pre-normalization check in HIR ty lowering for tuples & non-repeat arrays (as contrasted with the semantic `ConstArgHasType` check used for other expr kinds).

Reproducers:

```rs
#![feature(min_generic_const_args)]
#![expect(incomplete_features)]

type const T0: _ = ();
```

```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for constants
 --> file.rs:4:16
  |
4 | type const T0: _ = ();
  |                ^ not allowed in type signatures

error: expected `{type error}`, found const tuple
 --> file.rs:4:20
  |
4 | type const T0: _ = ();
  |                    ^^
```

```rs
#![feature(min_generic_const_args)]
#![expect(incomplete_features)]

type const T1 = [0];
```

```
error: missing type for `const` item
 --> file.rs:4:14
  |
4 | type const T1 = [0];
  |              ^
  |
help: provide a type for the item
  |
4 | type const T1: <type> = [0];
  |              ++++++++

error: expected `{type error}`, found const array
 --> file.rs:4:17
  |
4 | type const T1 = [0];
  |                 ^^^
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
@@ -2416,11 +2416,15 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     ) -> Const<'tcx> {
         let tcx = self.tcx();
 
-        let ty::Array(elem_ty, _) = ty.kind() else {
-            let e = tcx
-                .dcx()
-                .span_err(array_expr.span, format!("expected `{}`, found const array", ty));
-            return Const::new_error(tcx, e);
+        let elem_ty = match ty.kind() {
+            ty::Array(elem_ty, _) => elem_ty,
+            ty::Error(e) => return Const::new_error(tcx, *e),
+            _ => {
+                let e = tcx
+                    .dcx()
+                    .span_err(array_expr.span, format!("expected `{}`, found const array", ty));
+                return Const::new_error(tcx, e);
+            }
         };
 
         let elems = array_expr
@@ -2539,9 +2543,13 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     ) -> Const<'tcx> {
         let tcx = self.tcx();
 
-        let ty::Tuple(tys) = ty.kind() else {
-            let e = tcx.dcx().span_err(span, format!("expected `{}`, found const tuple", ty));
-            return Const::new_error(tcx, e);
+        let tys = match ty.kind() {
+            ty::Tuple(tys) => tys,
+            ty::Error(e) => return Const::new_error(tcx, *e),
+            _ => {
+                let e = tcx.dcx().span_err(span, format!("expected `{}`, found const tuple", ty));
+                return Const::new_error(tcx, e);
+            }
         };
 
         let exprs = exprs

--- a/tests/ui/const-generics/mgca/syntactic-type-mismatch.rs
+++ b/tests/ui/const-generics/mgca/syntactic-type-mismatch.rs
@@ -1,0 +1,13 @@
+// This test ensures proper diagnostics emission during HIR ty lowering
+// See https://github.com/rust-lang/rust/issues/153254
+
+#![feature(min_generic_const_args)]
+#![expect(incomplete_features)]
+
+type const T0: _ = ();
+//~^ ERROR: the placeholder `_` is not allowed within types on item signatures for constants [E0121]
+
+type const T1 = [0];
+//~^ ERROR: missing type for `const` item
+
+fn main() {}

--- a/tests/ui/const-generics/mgca/syntactic-type-mismatch.stderr
+++ b/tests/ui/const-generics/mgca/syntactic-type-mismatch.stderr
@@ -1,0 +1,20 @@
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for constants
+  --> $DIR/syntactic-type-mismatch.rs:7:16
+   |
+LL | type const T0: _ = ();
+   |                ^ not allowed in type signatures
+
+error: missing type for `const` item
+  --> $DIR/syntactic-type-mismatch.rs:10:14
+   |
+LL | type const T1 = [0];
+   |              ^
+   |
+help: provide a type for the item
+   |
+LL | type const T1: <type> = [0];
+   |              ++++++++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0121`.

